File: ciphers/vigenere_cipher.go
package main

import (
	"fmt"
	"strings"
	"unicode"
)

func translateMessage(key, message string, encrypt bool) string {
	letters := "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	translated := ""
	keyIndex := 0
	key = strings.ToUpper(key)

	for _, symbol := range message {
		upperSymbol := unicode.ToUpper(symbol)
		index := strings.IndexRune(letters, upperSymbol)

		if index != -1 {
			num := index
			keyVal := strings.IndexByte(letters, key[keyIndex])

			if encrypt {
				num += keyVal
			} else {
				num -= keyVal
			}

			if num < 0 {
				num += 26
			} else {
				num %= 26
			}

			if unicode.IsUpper(symbol) {
				translated += string(letters[num])
			} else {
				translated += strings.ToLower(string(letters[num]))
			}

			keyIndex++
			if keyIndex == len(key) {
				keyIndex = 0
			}
		} else {
			translated += string(symbol)
		}
	}
	return translated
}

func main() {
	text := "This is Harshil Darji from Dharmaj."
	key := "HDarji"

	encrypted := translateMessage(key, text, true)
	fmt.Println("Encrypted:", encrypted)
	decrypted := translateMessage(key, encrypted, false)
	fmt.Println("Decrypted:", decrypted)
}

File: ciphers/vigenere_cipher.cpp
#include <cctype>
#include <iostream>
#include <string>
#include <vector>

std::string translateMessage(std::string key, std::string message,
                             bool encrypt) {
  std::string translated = "";
  int keyIndex = 0;
  std::string letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

  for (int i = 0; i < key.length(); i++) {
    key[i] = toupper(key[i]);
  }

  for (char symbol : message) {
    if (isalpha(symbol)) {
      int num = -1;
      char upperSymbol = toupper(symbol);
      for (int i = 0; i < 26; i++) {
        if (letters[i] == upperSymbol) {
          num = i;
          break;
        }
      }

      int keyVal = -1;
      for (int i = 0; i < 26; i++) {
        if (letters[i] == key[keyIndex]) {
          keyVal = i;
          break;
        }
      }

      if (encrypt) {
        num += keyVal;
      } else {
        num -= keyVal;
      }

      // Handle negative wrap around manually for safety or just use correct
      // logic In Python % handles negatives well, in C++ % can return negative.
      if (num < 0) {
        num += 26;
      } else {
        num %= 26;
      }

      if (isupper(symbol)) {
        translated += letters[num];
      } else {
        translated += tolower(letters[num]);
      }

      keyIndex++;
      if (keyIndex == key.length()) {
        keyIndex = 0;
      }
    } else {
      translated += symbol;
    }
  }
  return translated;
}

int main() {
  std::string text = "This is Harshil Darji from Dharmaj.";
  std::string key = "HDarji";

  std::string encrypted = translateMessage(key, text, true);
  std::cout << "Encrypted: " << encrypted << std::endl;
  std::string decrypted = translateMessage(key, encrypted, false);
  std::cout << "Decrypted: " << decrypted << std::endl;
  return 0;
}
